csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3620 @@
+            intersection_exact_loop_graph.disjointIdSet(loop_id);
+        TORCH_INTERNAL_ASSERT(iel_set_pair.second);
+        auto iel_group = iel_set_pair.first;
+        auto iel_promo_it = iel_promotion_map.find(iel_group);
+        if (iel_promo_it == iel_promotion_map.end()) {
+          auto promo_id_exact_it =
+              idGraph(IdMappingMode::EXACT).disjointIdSet(loop_id);
+          TORCH_INTERNAL_ASSERT(promo_id_exact_it.second);
+          exact_promoted_terminal_ids.push_back(
+              std::make_pair(promo_id_exact_it.first, loop_id));
+        } else {
+          auto promo_id_exact_it =
+              idGraph(IdMappingMode::EXACT).disjointIdSet(iel_promo_it->second);
+          TORCH_INTERNAL_ASSERT(promo_id_exact_it.second);
+          exact_promoted_terminal_ids.push_back(
+              std::make_pair(promo_id_exact_it.first, iel_promo_it->second));
+        }
+      }
+    }
+
+    // All exact groups with iter domains in this loop group
+    IdGroups exact_groups;
+    for (auto loop_id : *loop_group) {
+      auto exact_set_pair =
+          idGraph(IdMappingMode::EXACT).disjointIdSet(loop_id);
+      TORCH_INTERNAL_ASSERT(exact_set_pair.second);
+      exact_groups.pushBack(exact_set_pair.first);
+    }
+
+    // All exact groups covered by all iter domains in this loop group
+    IdGroups loop_group_covered_ids;
+    for (auto exact_group : exact_groups) {
+      auto covered_it = exact_covered_ids.find(exact_group);
+      TORCH_INTERNAL_ASSERT(covered_it != exact_covered_ids.end());
+      loop_group_covered_ids.pushBack(covered_it->second);
+    }
+
+    IterDomain* loop_promotion_id = nullptr;
+
+    for (auto entry : exact_promoted_terminal_ids) {
+      auto terminal_id_group = entry.first;
+      auto terminal_id = entry.second;
+      auto covered_it = exact_covered_ids.find(terminal_id_group);
+      TORCH_INTERNAL_ASSERT(covered_it != exact_covered_ids.end());
+      if (loop_group_covered_ids.subtract(covered_it->second).size() == 0) {
+        loop_promotion_id = terminal_id;
+      }
+    }
+
+    if (loop_promotion_id == nullptr) {
+      std::stringstream err_msg;
+      err_msg
+          << "\n ERROR Loop promotion map build. Could not find promotion for loop group:\n  ";
+      err_msg << debug_print::idGroupStringShort(loop_group);
+      err_msg << "\nnone of the terminal iter domains of this group:\n  ";
+      for (auto entry : exact_promoted_terminal_ids) {
+        auto terminal_id_group = entry.first;
+        err_msg << "  " << debug_print::idGroupStringShort(terminal_id_group)
+                << std::endl;
+      }
+      err_msg << "iter domains in this group cover all id groups:\n";
+      for (auto covered_group : loop_group_covered_ids) {
+        err_msg << "  " << debug_print::idGroupStringShort(covered_group);
+      }
+      TORCH_INTERNAL_ASSERT(false, err_msg.str());
+    }
+
+    loop_graph_copy_promotion_map[loop_group] = loop_promotion_id;
+  }
+
```

Comment:
At this point, for each loop group, we have a mapping from a terminal ID that covers all IDs to its promoted ID. 

What is the second pass below going to do?

Reply:
That's not necessarily true. Only those that are inlined.

If a tensor view has a resolved broadcast domain inlined with a consumer, at this point that iter domain would be resolved correctly.

If that tensor view then has transformations after the promoted iter domain, any transformations after that would not be correctly replayed on top of that promoted iter domain.


The second pass is replaying those transformations.

`T4[i0, i1] = T3[b0, i1] + T1[i0, i1]`

If `T3[b0*i1//32//8, 32//8, 8]` and `T4[i0*i1//32, 32]` with inlining of T3 into T4 at position 1 then T3 at this point would have `i0*i1//32//8` promoted as its first iter domain, however, the history of `32//8` and `8` would still be based on `b0*i1`.

So yes, the inlined axes at this point are replayed and promoted correctly, the leaves are not.